src/globus_sdk/scopes/scope_definition.py

```diff
@@ -0,0 +1,323 @@
+"""
+This defines the Scope object and the scope parser.
+Because these components are mutually dependent, it's easiest if they're kept in a
+single module.
+"""
+import enum
+import typing as t
+import warnings
+
+from globus_sdk._types import ScopeCollectionType
+
+
+class ScopeParseError(ValueError):
+    pass
+
+
+def _iter_scope_collection(obj: ScopeCollectionType) -> t.Iterator[str]:
+    if isinstance(obj, str):
+        yield obj
+    elif isinstance(obj, Scope):
+        yield str(obj)
+    else:
+        for item in obj:
+            yield str(item)
+
+
+class ParseTokenType(enum.Enum):
+    # a string like 'urn:globus:auth:scopes:transfer.api.globus.org:all'
+    scope_string = enum.auto()
+    # the optional marker, '*'
+    opt_marker = enum.auto()
+    # '[' and ']'
+    lbracket = enum.auto()
+    rbracket = enum.auto()
+
+
+class ParseToken:
+    def __init__(self, value: str, token_type: ParseTokenType) -> None:
+        self.value = value
+        self.token_type = token_type
+
+
+def _tokenize(scope_string: str) -> t.List[ParseToken]:
+    tokens: t.List[ParseToken] = []
+    current_token: t.List[str] = []
+    for idx, c in enumerate(scope_string):
+        try:
+            peek: t.Optional[str] = scope_string[idx + 1]
+        except IndexError:
+            peek = None
+
+        if c in "[]* ":
+            if current_token:
+                tokens.append(
+                    ParseToken("".join(current_token), ParseTokenType.scope_string)
+                )
+                current_token = []
+
+            if c == "*":
+                if peek == " ":
+                    raise ScopeParseError("'*' must not be followed by a space")
```

Comment:
While researching whether `*[scope]` was acceptable syntax I noticed that [scopes may be plus-separated](https://docs.globus.org/api/auth/reference/#grant_request), too. Should that be addressed here?

Reply:
I believe that that's a misleading bit of documentation. Unless I'm mistaken, it's just referring to `+` as a style of url quoting.

Reply:
I wondered if that might be the case. I'll mark this resolved. Thanks!